### PR TITLE
Persist switcher cards

### PR DESCRIPTION
**Pull request by Parveshdhull**
status-go PR: https://github.com/status-im/status-go/pull/3049

### Testing:
PR handles storing the card when opening chat, community, etc. And deleting the card by pressing the close button.
Removing the card when deleting chat, blocking user, etc. will be implemented later. 

status : ready

### CODE REVIEW

File: src/status_im2/contexts/shell/events.cljs
Comment (Parveshdhull) [inline src/status_im2/contexts/shell/events.cljs]: extra field: https://github.com/status-im/status-go/pull/3049#discussion_r1062277109

File: src/status_im/data_store/switcher_cards.cljs
Comment (flexsurfer) [inline src/status_im/data_store/switcher_cards.cljs:36]: just wondering why this one in a separate ns ?

File: src/status_im2/contexts/shell/events.cljs
Comment (ilmotta) [reply src/status_im2/contexts/shell/events.cljs]: FYI: @Parveshdhull explained some fields will be renamed in status-go, so my comment is irrelevant now.

Outdated:

As I mentioned over there in the status-go PR, maybe renaming the key from `:extra` to `:metadata` could be more clear, because metadata is a word that usually implies some structure.

I would prefer if we could persist stuff in this `extra/metadata` field as serialized JSON. That way we always know what's stored from inspecting the DB, otherwise the value is too opaque. Of course, that means you'll need to call `JSON.parse` to deserialize, but that's trivial and worth it IMO.

File: src/status_im2/contexts/shell/events.cljs
Comment (ilmotta) [inline src/status_im2/contexts/shell/events.cljs]: You can replace the entire reduce processing with `status-im.utils.core/index-by`.

File: src/status_im2/contexts/shell/events.cljs
Comment (Parveshdhull) [reply src/status_im2/contexts/shell/events.cljs]: resolved in status-go. Thank you @ilmotta

File: src/status_im/data_store/switcher_cards.cljs
Comment (Parveshdhull) [reply src/status_im/data_store/switcher_cards.cljs:36]: fixed, thank you

File: src/status_im2/contexts/shell/events.cljs
Comment (Parveshdhull) [reply src/status_im2/contexts/shell/events.cljs]: now, also renaming keys. So index-by will not be sufficient

Comment (ilmotta) [reply src/status_im2/contexts/shell/events.cljs]: @Parveshdhull, you can still use `index-by` because the collection can be transformed outside the reducing function.

```clojure
(->> loaded-switcher-cards
     (map switcher-cards-store/<-rpc)
     (utils.core/index-by :card-id))
```

File: src/status_im/data_store/switcher_cards.cljs
Comment (ilmotta) [inline src/status_im/data_store/switcher_cards.cljs]: I believe we're following the Clojure Style Guide for certain namespace aliases. See the table here https://guide.clojure.style/#use-idiomatic-namespace-aliases

`clojure.set` should be aliased as `set`

File: src/status_im2/contexts/shell/events.cljs
Comment (ilmotta) [inline src/status_im2/contexts/shell/events.cljs]: Just `update` is sufficient instead of `update-in`.

**Review by ilmotta — APPROVED** [review]

**Review by siddarthkay — APPROVED** [review]

File: src/status_im2/contexts/shell/events.cljs
Comment (Parveshdhull) [reply src/status_im2/contexts/shell/events.cljs]: hi, thank you for suggestion, but only update is not working

File: src/status_im2/contexts/shell/events.cljs
Comment (Parveshdhull) [reply src/status_im2/contexts/shell/events.cljs]: done using `clojure.walk`, thank you

File: src/status_im2/contexts/shell/events.cljs
Comment (ilmotta) [reply src/status_im2/contexts/shell/events.cljs]: `update` without the square brackets around the key should work.

`(update db :shell/switcher-cards dissoc card-id)`

```clojure
(let [card-id "0x1"
      db      {:shell/switcher-cards {card-id {:card-data :anything}}}]
  (assert (= {:shell/switcher-cards {}}
             (update db :shell/switcher-cards dissoc card-id))))
```

Comment (Parveshdhull) [reply src/status_im2/contexts/shell/events.cljs]: oh, missed that, working now. Thank you

File: src/status_im/data_store/switcher_cards.cljs
Comment (ilmotta) [inline src/status_im/data_store/switcher_cards.cljs:20]: Just a quick thing I didn't realize before @Parveshdhull. I had the impression the data store directory should know nothing about re-frame. This layer should only care about the domain model, often using only pure functions.

Event handlers should be added to another layer of the stack IMO, in `events.cljs` files somewhere. I think this is rather important.

Comment (ilmotta) [reply src/status_im/data_store/switcher_cards.cljs:20]: I see there are other namespaces in `data_store/` defining event handlers, but this is a mistake IMO, especially now that we have new guidelines.

So event handlers should be added to `events.cljs` files in their corresponding `status-im2/contexts/*` directory, or to `status-im2/common` if they're really general.

**Comment by qoqobolo** [comment]
Hey @Parveshdhull, thanx for PR, works great!

Two things here:

1. Q: Is it expected, that a card is not persisted on the first opening of a chat?

I.e. if you create a new group chat or a community and go back to the home screen and then press `Jump to`, this card won't be there. For the chat card to appear on the screen, you need to open that chat again from the chat screen.
It behaves this way in develop as well, I'm just wondering if we want to fix it.

https://user-images.githubusercontent.com/67952253/210816634-9d3a833a-9930-42c0-88ad-6a73b4ca242a.mp4

(asked the same question [here ](https://github.com/status-im/status-mobile/pull/14717#issuecomment-1372360608), please answer wherever you want :))

2. In this PR I can't test navigation from the chat view due to this issue https://github.com/status-im/status-mobile/issues/14668.

But it has already been fixed in #14717.
I don't want to block this PR due to 14668, so actually it can be merged.
Or we can wait until 14717 is merges, then you rebase your PR and I can test that flow.
WDYT?

**Comment by Parveshdhull** [comment]
Hi @qoqobolo,
Thank you very much for testing the PR.

Actually, adding of switcher card is not handled by the jump-to button. Jump-to just closes the corresponding screen, so that we can see the shell screen and access cards.

Adding cards is handled by navigation, and currently, only the home screen is using navigation2 calls. We have to use it at several other places like communities list (as shown in the video), or new 1-1 to chat or starting a chat from a contact. I will work on that in a separate PR. 

Also, about Q2, we can skip that, as I mentioned jump-to doesn't have anything to do with cards.

Will that be ok?

**Comment by qoqobolo** [comment]
@Parveshdhull perfectly clear now, thank you!

>Also, about Q2, we can skip that, as I mentioned jump-to doesn't have anything to do with cards.
Will that be ok?

Yep, PR can be merged.

**Review by flexsurfer — APPROVED** [review]

**Pull request merged**
